spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/FirestoreTemplate.java

```diff
@@ -115,4 +137,43 @@ public FirestoreTemplate(FirestoreStub firestore, String parent) {
 		return ObservableReactiveUtil.<RunQueryResponse>streamingCall(obs -> this.firestore.runQuery(request, obs))
 				.filter(RunQueryResponse::hasDocument).map(RunQueryResponse::getDocument);
 	}
+
+	private StreamObserver<WriteRequest> openWriteStream(StreamObserver<WriteResponse> obs) {
+		WriteRequest openStreamRequest = WriteRequest.newBuilder().setDatabase(this.databasePath).build();
+		StreamObserver<WriteRequest> requestStreamObserver = this.firestore.write(obs);
+		requestStreamObserver.onNext(openStreamRequest);
+		return requestStreamObserver;
+	}
+
+	private <T> Mono<WriteRequest> writeEntityStream(T entity, Flux<WriteResponse> responses) {
+		Mono<WriteResponse> firstResponse = responses.next();
+		return firstResponse.map(
+			initialResponse -> buildWriteRequest(initialResponse.getStreamId(), initialResponse.getStreamToken(), entity));
```

Comment:
`buildResourceName()` should not be done for every entity; it should be moved up to the top level where entity type is known. But it was awkward to drag the name through several method signatures, and this will get refactored again anyway.